core/src/main/java/com/linecorp/armeria/client/endpoint/GradualStartEndpointSelectionStrategy.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.endpoint;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.endpoint.WeightedRoundRobinStrategy.EndpointsAndWeights;
+import com.linecorp.armeria.common.CommonPools;
+
+/**
+ * A weighted round-robin strategy that ramps up the weight gradually from zero to its normal weight value.
+ */
+public final class GradualStartEndpointSelectionStrategy implements EndpointSelectionStrategy {
+    private static final int DEFAULT_NUMBER_OF_STEPS = 20;
+
+    /**
+     * Creates a {@link GradualStartEndpointSelectionStrategyBuilder}.
+     */
+    public static GradualStartEndpointSelectionStrategyBuilder builder() {
+        return new GradualStartEndpointSelectionStrategyBuilder();
+    }
+
+    /**
+     * Creates a {@link GradualStartEndpointSelectionStrategy}.
+     */
+    public static EndpointSelectionStrategy of(Duration slowStartInterval) {
```

Comment:
I think total duration is better as a parameter in this case because the number of steps is not clear here.